### PR TITLE
Update zero tracking

### PR DESCRIPTION
**Pull request by FabioLuporini**
- Quite a lot of refactoring
- Improves tracking of block-sparse arrays (like those arising when using VFS), plus bug fixes. This tracking is on only with optimization level O3
- Fixes padding when offsets are presents (e.g. A[i+k])

This should land together with https://bitbucket.org/mapdes/ffc/pull-request/80/update-to-new-coffee-interface/diff

### CODE REVIEW

File: coffee/plan.py [hunk 1]
```diff
@@ -399,6 +384,7 @@ def _generate_cpu_code(self, kernel, **kwargs):
 
             # Post processing of the AST ensures higher-quality code
             postprocess(kernel)
+        print self.ast
```
Comment (kynan) [inline coffee/plan.py]: Leftover from debugging?

**Comment by kynan** [comment]
From a very high-level scan this looks sane.

File: coffee/plan.py (hunk 1 above)
Comment (FabioLuporini) [inline coffee/plan.py]: woooo yes! thanks!

**Comment by miklos1** [comment]
Why is it not in master yet?

**Comment by FabioLuporini** [comment]
needs rebase onto latest master, since the latest PR changed bits in the the visitor infrastructure.

I'm planning to do this on Wed

**Comment by FabioLuporini** [comment]
Rebased on master.
Also cherry-picked a few comments from enhance-rewriter-mode3 that fix one bug with expansion. 
Seem to pass all tests at both O2 and O3 levels (except from test_vfs_... when running the whole test suite)
@miklos1  @wence-  have a look when you have time and then let's merge it

**Comment by miklos1** [comment]
Pleease, pay attention to Travis, you have flake8 problem.

**Comment by FabioLuporini** [comment]
Despite updating flake8 on my machine, I still don't see this on local ... anyway, fixed it

File: coffee/rewriter.py [hunk 2]
```diff
@@ -886,14 +895,15 @@ def _factorize(self, node, parent):
             # Example: replace (a*b)+(a*b) with 2*(a*b)
             self._simplify_sum(terms)
             # Finally try to factorize some of the operands composing the operation
-            factorized = {}
+            factorized = OrderedDict()
             for t in terms:
                 operand = set([t.operands_ast]) if t.operands else set()
-                factor = set([t.factors_ast]) if t.factors else set()
+                factor = set([t.factors_ast]) if t.factors else set([Symbol(1.0)])
```
Comment (miklos1) [inline coffee/rewriter.py]: Why is a number a Symbol?

Comment (FabioLuporini) [inline coffee/rewriter.py]: Because it's part of the generated code

**Pull request merged**